### DISCUSSION TITLE
Development Environment

### DISCUSSION THREAD

**Discussion by pstaabp** [openwebwork/webwork2#1173] [Q&A]
First, just noticed that GitHub now has these discussions instead of just issues.  Thanks for turning this on @mgage.

I was wondering for the developers out there, how you set up an environment for development.  Is anyone using Docker to do this?  I find it incredibly painful to do any development with Docker. 

To start things out, For the past 10 years or so, I've been on a Mac and have been running everything natively on MacOS, which generally works out ok.  [Here's](https://github.com/pstaabp/webwork2/wiki/Developing-Webwork-on-MacOS) the notes that I've developed for myself as I often need to rebuild this.    One issue that often pops up is that when MacOS upgrades there is often problems getting started again.  I just did an upgrade recently to BigSur, so am working through a few things here. 

Even though once I get things generally working well on my machine, I'm open to other ways of doing things.

**Comment by drgrice1** [comment]
I run Ubuntu for my primary operating system, and I just install webwork directly on my local computer.  Then I can make changes to the code, use git to switch branches, etc.  Just remember to restart apache2 when changes are made.

For those running MacOSX or MSWindows it is probably more complicated.
